src/Tvheadend.cpp

```diff
@@ -223,6 +224,50 @@ bool CTvheadend::HasStreamingProfile(const std::string& streamingProfile) const
                       { return profile.GetName() == streamingProfile; }) != m_profiles.cend();
 }
 
+/* **************************************************************************
+ * Providers
+ * *************************************************************************/
+
+PVR_ERROR CTvheadend::GetProvidersAmount(int& amount)
+{
+  if (!m_asyncState.WaitForState(ASYNC_DVR))
+    return PVR_ERROR_FAILED;
+
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+  amount = m_providers.size();
+  return PVR_ERROR_NO_ERROR;
+}
+
+PVR_ERROR CTvheadend::GetProviders(kodi::addon::PVRProvidersResultSet& results)
+{
+  std::vector<kodi::addon::PVRProvider> providers;
+  {
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
+
+    providers.reserve(m_providers.size());
+    for (const auto& entry : m_providers)
+    {
+      kodi::addon::PVRProvider provider;
+      provider.SetUniqueId(entry.second.GetId());
+      provider.SetName(entry.second.GetName());
+      //SetType
+      //SetIconPath
+      //SetCountries
+      //SetLanguages
+
+      providers.emplace_back(std::move(provider));
```

Comment:
Why not `std::move()` directly to `results`?

Reply:
Results is not a simple container. `Add` actually cllas back to Kodi to immediately deliver the result. We use the concept "collect results mutex ptrotected, then release mutex and deliver results to Kodi" everywhere in the add-on, also for channels, groups, ...

Reply:
Right :+1: